### PR TITLE
fixes #6216 - foreign key errors such as deleting a subnet that is associated with a host group and others

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by isratrade** [comment]
I suggest to have Ori review other places foreign keys are missing `before_destroy EnsureNotUsedBy.new`

**Comment by isratrade** [comment]
not sure why there were 2 katello test failures. re [test] ing

**Comment by GregSutcliffe** [comment]
Tests fine locally, but this seems like the sort of thing we could easily test with FactoryGirl - could you try to add tests so we don't hit this again?

**Comment by isratrade** [comment]
@GregSutcliffe, I added a `subnet` factory and two tests to check validation error is generated on `destroy` for `hosts` or `hostgroups`.

File: test/factories/subnet.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :subnet do
+    sequence(:name) {|n| "subnet#{n}" }
+    sequence(:network) {|n| "10.0.0.#{n}" }
+    sequence(:mask) {|n| "255.255.255.#{n}" }
```
Comment (GregSutcliffe) [inline test/factories/subnet.rb]: From a networking perspective this isn't quite right, as "10.0.0.3/255.255.255.3" isn't valid (or is at least a very weird subnet). It should probably be something like:

```
sequence(:network) {|n| "10.0.#{n}.0" }
mask '255.255.255.0'
```

**Comment by isratrade** [comment]
@GregSutcliffe, thanks. re-committed with your suggested changes.

**Comment by GregSutcliffe** [comment]
:+1: merge at will

**Comment by domcleal** [comment]
Please don't merge until the review for other instances is done.

**Comment by domcleal** [comment]
The commit message references the wrong issue, should be #6216: http://projects.theforeman.org/issues/6216

File: app/models/compute_profile.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@ class ComputeProfile < ActiveRecord::Base
   audited
   has_associated_audits
 
-  before_destroy EnsureNotUsedBy.new(:hostgroups)
+  before_destroy EnsureNotUsedBy.new(:hosts, :hostgroups)
```
Comment (isratrade) [inline app/models/compute_profile.rb]: there is a FK in the db, but I'm wondering if there should be. After a host is created, the compute_profile is no longer used and not displayed on the host edit page.

File: app/models/domain.rb [hunk 3]
```diff
@@ -8,7 +8,7 @@ class Domain < ActiveRecord::Base
   has_many_hosts
   has_many :hostgroups
   #order matters! see https://github.com/rails/rails/issues/670
-  before_destroy EnsureNotUsedBy.new(:hosts, :hostgroups, :subnets)
+  before_destroy EnsureNotUsedBy.new(:hosts, :hostgroups)
```
Comment (isratrade) [inline app/models/domain.rb]: no need for `:subnets` since there is a ':dependant => :destory' on the join table below

File: app/models/environment.rb [hunk 4]
```diff
@@ -2,17 +2,17 @@ class Environment < ActiveRecord::Base
...
   has_many :puppetclasses, :through => :environment_classes, :uniq => true
   has_many_hosts
   has_many :hostgroups
-  has_many :trends, :as => :trendable, :class_name => "ForemanTrend"
+  has_many :trends, :as => :trendable, :class_name => "ForemanTrend", :dependent => :destroy
 
   validates :name, :uniqueness => true, :presence => true, :format => { :with => /\A[\w\d]+\Z/, :message => N_("is alphanumeric and cannot contain spaces") }
-  has_many :config_templates, :through => :template_combinations
   has_many :template_combinations, :dependent => :destroy
+  has_many :config_templates, :through => :template_combinations
```
Comment (isratrade) [inline app/models/environment.rb:15]: nitpick, just changed order, I usually prefer the join table relationship first.

File: app/models/image.rb [hunk 5]
```diff
@@ -3,6 +3,8 @@ class Image < ActiveRecord::Base
 
   audited :allow_mass_assignment => true
 
+  before_destroy EnsureNotUsedBy.new(:hosts)
```
Comment (isratrade) [inline app/models/image.rb]: not sure about this, do you agree?

File: app/models/model.rb [hunk 6]
```diff
@@ -1,9 +1,9 @@
 class Model < ActiveRecord::Base
   include Authorizable
 
-  has_many_hosts
-  has_many :trends, :as => :trendable, :class_name => "ForemanTrend"
   before_destroy EnsureNotUsedBy.new(:hosts)
```
Comment (isratrade) [inline app/models/model.rb:4]: `before_destroy` is supposed to be before the `has_many` definitions

File: app/models/user.rb [hunk 7]
```diff
@@ -21,7 +21,7 @@ class User < ActiveRecord::Base
   has_many :direct_hosts,      :class_name => 'Host',    :as => :owner
   has_many :usergroup_member,  :dependent => :destroy,   :as => :member
   has_many :user_roles,        :dependent => :destroy, :foreign_key => 'owner_id', :conditions => {:owner_type => self.to_s}
-  has_many :user_hostgroups,   :dependent => :destroy
+  has_many :user_hostgroups
```
Comment (isratrade) [inline app/models/user.rb:24]: I added `:hostgroups` and removed `:dependent => :destroy`

File: app/models/compute_profile.rb (hunk 2 above)
Comment (ohadlevy) [inline app/models/compute_profile.rb]: i tend to agree.

File: app/models/domain.rb (hunk 3 above)
Comment (ohadlevy) [inline app/models/domain.rb]: i'm not sure if thats right, as subnets <-> domains have many to many relationships.

what if i deleted a domain by mistake, it would trigger deleting all of its subnets? I think you still want to show an error message acknowledging you want to delete all subnets (and ensure they are not used by other doimains?)

File: app/models/compute_resource.rb [hunk 8]
```diff
@@ -18,7 +18,7 @@ class ComputeResource < ActiveRecord::Base
 
   audited :except => [:password, :attrs], :allow_mass_assignment => true
   serialize :attrs, Hash
-  has_many :trends, :as => :trendable, :class_name => "ForemanTrend"
+  has_many :trends, :as => :trendable, :class_name => "ForemanTrend", :dependent => :destroy
```
Comment (ohadlevy) [inline app/models/compute_resource.rb]: this is going to be painful if there a lot of trendcounters, ideally should add a todo here do delete in bg when possible.

**Comment by ohadlevy** [comment]
[test]

File: app/models/compute_profile.rb (hunk 2 above)
Comment (isratrade) [inline app/models/compute_profile.rb]: @ohadlevy, ok I will delete FK from db and remove this change

File: app/models/compute_resource.rb (hunk 8 above)
Comment (isratrade) [inline app/models/compute_resource.rb]: @ohadlevy, ok I will remove the `dependant => :destroy` on `has_many :trends`

File: app/models/domain.rb (hunk 3 above)
Comment (isratrade) [inline app/models/domain.rb]: @ohadlevy, is this also true in the opposite direction to not delete a subnet if there are associated domains?

Comment (ohadlevy) [inline app/models/domain.rb]: i think so.

File: app/models/usergroup.rb [hunk 9]
```diff
@@ -2,6 +2,8 @@ class Usergroup < ActiveRecord::Base
   audited :allow_mass_assignment => true
   include Authorizable
 
+  before_destroy EnsureNotUsedBy.new(:hosts)
```
Comment (isratrade) [inline app/models/usergroup.rb:5]: when I moved `before_destroy EnsureNotUsedBy.new(:hosts, :usergroups)` above the `has_many` definitions, it caused a usergroup_test failure on  "can be destroyed when in use by another usergroup, it removes association automatically". So, I removed `:usergroups` from the `EnsureNotUsedBy.new` parameters

File: app/models/image.rb (hunk 5 above)
Comment (domcleal) [inline app/models/image.rb]: Yeah, I don't think we should enforce it.. http://projects.theforeman.org/issues/4416

**Comment by domcleal** [comment]
Thanks, looks good - one comment above in response to a question.

File: app/models/compute_profile.rb [hunk 10]
```diff
@@ -9,7 +9,7 @@ class ComputeProfile < ActiveRecord::Base
   before_destroy EnsureNotUsedBy.new(:hostgroups)
   has_many :compute_attributes, :dependent => :destroy
   has_many :compute_resources, :through => :compute_attributes
-  has_many_hosts
+  has_many_hosts :dependent => :nullify
```
Comment (isratrade) [inline app/models/compute_profile.rb:12]: I deleted the db migration that was added to remove the FK constraint on hosts.compute_profile_id and added `:dependent => :nullify`

File: app/models/image.rb [hunk 11]
```diff
@@ -7,7 +7,7 @@ class Image < ActiveRecord::Base
   belongs_to :compute_resource
   belongs_to :architecture
 
-  has_many_hosts
+  has_many_hosts :dependent => :nullify
```
Comment (isratrade) [inline app/models/image.rb:10]: @domcleal, this fixes http://projects.theforeman.org/issues/4416
Do you want me to make a separate PR just for this?

Comment (domcleal) [inline app/models/image.rb:10]: no, don't worry, I'll link and close them together

**Comment by domcleal** [comment]
Thanks @isratrade, merged as e0d618e for Foreman 1.5.2.
